File: oc/AnimationDemo/AnimationDemo/MainViewController.m
//
//  MainViewController.m
//  AnimationDemo
//
//  Created by dfang on 2018-3-27.
//  Copyright © 2018年 east. All rights reserved.
//

#import "MainViewController.h"
#import "BasicAnimationViewController.h"
#import "KeyFrameAnimationViewController.h"
#import "TransitionAnimationViewController.h"
#import "GroupAnimationViewController.h"
#import "SummaryAnimationViewController.h"

@interface MainViewController ()

@end

@implementation MainViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    // method 1: assign viewcontrollers
    [self addViewControllers];
    
    // method2: addchildviewcontroller
    [self addChildViewController:[[BasicAnimationViewController alloc] init] withTitle:@"基本动画" andImageName:@"TabBar1"];
    [self addChildViewController:[[KeyFrameAnimationViewController alloc] init] withTitle:@"帧动画" andImageName:@"TabBar2"];
    [self addChildViewController:[[TransitionAnimationViewController alloc] init] withTitle:@"转场动画" andImageName:@"TabBar3"];
    [self addChildViewController:[[GroupAnimationViewController alloc] init] withTitle:@"组动画" andImageName:@"TabBar4"];
    [self addChildViewController:[[SummaryAnimationViewController alloc] init] withTitle:@"动画总结" andImageName:@"TabBar5"];
}

// assign viewcontrollers
- (void)addViewControllers {
    BasicAnimationViewController *basicAni = [[BasicAnimationViewController alloc] init];
    basicAni.tabBarItem.title = @"基本动画";
    basicAni.navigationItem.title = @"基本动画演示";
    basicAni.tabBarItem.imageInsets = UIEdgeInsetsMake(6, 0, -6, 0);
    basicAni.tabBarItem.image = [[UIImage imageNamed:@"TabBar1"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    basicAni.tabBarItem.selectedImage = [[UIImage imageNamed:@"TabBar1Sel"] imageWithRenderingMode: UIImageRenderingModeAlwaysOriginal];
    KeyFrameAnimationViewController *keyframAni = [[KeyFrameAnimationViewController alloc] init];
    keyframAni.tabBarItem.title = @"帧动画";
    keyframAni.tabBarItem.imageInsets = UIEdgeInsetsMake(6, 0, -6, 0);
    keyframAni.tabBarItem.image = [[UIImage imageNamed:@"TabBar2"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    keyframAni.tabBarItem.selectedImage = [[UIImage imageNamed:@"TabBar2Sel"] imageWithRenderingMode: UIImageRenderingModeAlwaysOriginal];
    TransitionAnimationViewController *transiAni = [[TransitionAnimationViewController alloc] init];
    transiAni.tabBarItem.title = @"转场动画";
    transiAni.tabBarItem.imageInsets = UIEdgeInsetsMake(6, 0, -6, 0);
    transiAni.tabBarItem.image = [[UIImage imageNamed:@"TabBar3"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    transiAni.tabBarItem.selectedImage = [[UIImage imageNamed:@"TabBar3Sel"] imageWithRenderingMode: UIImageRenderingModeAlwaysOriginal];
    GroupAnimationViewController *groupAni = [[GroupAnimationViewController alloc] init];
    groupAni.tabBarItem.title = @"组动画";
    groupAni.tabBarItem.imageInsets = UIEdgeInsetsMake(6, 0, -6, 0);
    groupAni.tabBarItem.image = [[UIImage imageNamed:@"TabBar4"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    groupAni.tabBarItem.selectedImage = [[UIImage imageNamed:@"TabBar4Sel"] imageWithRenderingMode: UIImageRenderingModeAlwaysOriginal];
    SummaryAnimationViewController *summaryAni = [[SummaryAnimationViewController alloc] init];
    summaryAni.tabBarItem.title = @"动画总结";
    summaryAni.tabBarItem.imageInsets = UIEdgeInsetsMake(6, 0, -6, 0);
    summaryAni.tabBarItem.image = [[UIImage imageNamed:@"TabBar5"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    summaryAni.tabBarItem.selectedImage = [[UIImage imageNamed:@"TabBar5Sel"] imageWithRenderingMode: UIImageRenderingModeAlwaysOriginal];
    
    self.viewControllers = @[[[UINavigationController alloc] initWithRootViewController:basicAni],
                             [[UINavigationController alloc] initWithRootViewController:keyframAni],
                             [[UINavigationController alloc] initWithRootViewController:transiAni],
                             [[UINavigationController alloc] initWithRootViewController:groupAni],
                             [[UINavigationController alloc] initWithRootViewController:summaryAni]
                             ];
}

// add childviewcontroller
- (void)addChildViewController:(UIViewController *)childController withTitle: (NSString *)title andImageName: (NSString *)imageName {
    childController.tabBarItem.title = title;
    childController.tabBarItem.imageInsets = UIEdgeInsetsMake(6, 0, -6, 0);
    childController.tabBarItem.image = [[UIImage imageNamed:imageName] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    childController.tabBarItem.selectedImage = [[UIImage imageNamed:[NSString stringWithFormat:@"%@Sel", imageName]] imageWithRenderingMode: UIImageRenderingModeAlwaysOriginal];
    
    [self addChildViewController:[[UINavigationController alloc] initWithRootViewController:childController]];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: swift/AnimationDemo/AnimationDemo/MainViewController.swift
//
//  MainViewController.swift
//  AnimationDemo
//
//  Created by dfang on 2018-3-19.
//  Copyright © 2018年 east. All rights reserved.
//

import UIKit

class MainViewController: UITabBarController {
    
    override func viewDidLoad() {
        view.backgroundColor = UIColor.white
        tabBar.tintColor = UIColor.orange
        creatSubViewControllers()
    }
    
    func creatSubViewControllers() {
        // 先创建好所有的vc再一次添加到viewcontrollers中
        let basicVC = BasicAnimationViewController.init()
        addChildViewController(childController: basicVC, title: "基本动画", imageName: "TabBar1")
        let kfVC = KeyFrameViewController.init()
        addChildViewController(childController: kfVC, title: "帧动画", imageName: "TabBar2")
        let transitionVC = TransitionViewController.init()
        addChildViewController(childController: transitionVC, title: "转场动画", imageName: "TabBar3")
        let agVC = AnimationGroupViewController.init()
        addChildViewController(childController: agVC, title: "组动画", imageName: "TabBar4")
        let summaryVC = SummaryViewController.init()
        addChildViewController(childController: summaryVC, title: "动画总结", imageName: "TabBar5")
        
        // If you change the value of this property at runtime,
        // the tab bar controller removes all of the old view controllers before installing the new ones.
        // The tab bar items for the new view controllers are displayed immediately and are not animated into position.
        viewControllers = [UINavigationController.init(rootViewController: basicVC),
                           UINavigationController.init(rootViewController: kfVC),
                           UINavigationController.init(rootViewController: transitionVC),
                           UINavigationController.init(rootViewController: agVC),
                           UINavigationController.init(rootViewController: summaryVC)]
        
        // 更改more对应tableview样式，不推荐，也许某天apple将topViewController?.view更改为非tableview
        let topView = moreNavigationController.topViewController?.view
        if (topView?.isKind(of: UITableView.self))! {
            let tableView:UITableView = topView as! UITableView

            // 设置datasource设置代理，在代理方法中处理
//            tableView.dataSource = self
//            tableView.delegate = self
            
            tableView.rowHeight = 100
            tableView.tableFooterView = UIView.init(frame: CGRect.zero)
        }
        
        // 2 创建一个VC就添加到viewcontrollers中
//        addChildViewController(childController: BasicAnimationViewController(), title:"基本动画", imageName:"TabBar1")
//        addChildViewController(childController: KeyFrameViewController(), title:"帧动画", imageName: "TabBar2")
//        addChildViewController(childController: TransitionViewController(), title: "转场动画", imageName: "TabBar3")
//        addChildViewController(childController: AnimationGroupViewController(), title: "组动画", imageName: "TabBar4")
//        addChildViewController(childController: SummaryViewController(), title: "动画总结", imageName: "TabBar5")
    }
    
    // ios7之后设置tabBarItem的image需要渲染否则图片不会显示，默认图片会向上偏移，需设置tabBarItem的imageInsets
    func addChildViewController(childController: UIViewController,title:String,imageName:String) {
        childController.tabBarItem.title = title
        childController.tabBarItem.imageInsets = UIEdgeInsetsMake(6, 0, -6, 0)
        childController.tabBarItem.image = UIImage.init(named: imageName)?.withRenderingMode(UIImageRenderingMode.alwaysOriginal)
        childController.tabBarItem.selectedImage = UIImage(named: imageName + "Sel")?.withRenderingMode(UIImageRenderingMode.alwaysOriginal)
        
        // 2.1
//        if (viewControllers == nil) {
//            viewControllers = [UINavigationController.init(rootViewController: childController)]
//        } else {
//            viewControllers!.append(UINavigationController.init(rootViewController: childController))
//        }
        
        // 2.2 超过5个后，后面的不显示
//        addChildViewController(UINavigationController.init(rootViewController: childController))
    }
}
